### ISSUE TITLE
Check reproduced grasp evaluation result

### ISSUE THREAD

**Issue by ShaoxiongYao** [4DVLab/DexGrasp-Anything#29]
Thank you for providing the great codebase! I am trying to reproduce the results in the paper, in particular I use the pre-trained DexGraspNet checkpoint to evaluate on the DexGraspNet dataset. I used the following command to generate the results:

```
bash scripts/grasp_gen_ur/sample.sh ${exp_dir} [OPT]
```

The code runs successfully and produces html visualization in eval folder, but I realized there are frequent cases where the model predicts infeasible grasp with deep penetration with the object, as shown in the following video:

https://github.com/user-attachments/assets/b52fd799-8dfb-4859-9e59-abab6860a577

I wonder if this is expected result or you may have some suggestions on potential fixes. Thank you for your help in advance!

**Comment by ymzhong66** [comment]
Yes, a small number of cases like this is expected.
If you want to minimize such penetration cases, you may consider adjusting the penetration-guidance weight accordingly.

**Comment by ShaoxiongYao** [comment]
Thank you for your response! I took a closer look at the diffusion process where grasp poses are generated. Specifically, I added a few lines at 

https://github.com/4DVLab/DexGrasp-Anything/blob/a242f50d06358117ab9c8b0f958e79e3e8df1dd6/models/optimizer/grasp_with_object.py#L67C9-L69C96

to save `hand_pcd` and `obj_pcd` for visualization purposes:
```python
hand_pcd = self.hand_model.get_surface_points(q=x).to(dtype=torch.float32)
normal = data['normal'].clone().detach().requires_grad_(True).to(self.device)
obj_pcd = rearrange(obj_pcd, '(b n) c -> b n c', b=self._BATCH_SIZE, n=normal.shape[1])
```

I created a short animation to visualize the grasp generation process, particularly tracking how the hand point cloud evolves over time. However, I noticed that the hand configuration makes only minimal changes during the diffusion process. This seems to mean the diffusion is not doing too much update, where I would expect more significant updates. I'm wondering if this might be because the visualization captures the state after only one iteration of the `p_sample` function, though I'm not certain about this.

Does this behavior align with your implementation, or am I misunderstanding something?

https://github.com/user-attachments/assets/7cd528f1-796a-4030-ac87-1fdd4d03d4a0

**Comment by ymzhong66** [comment]
Are you visualizing the \(x_0\) reconstructed from each \(x_t\) during the reverse diffusion process?  
I’ve visualized this behavior before: if you only look at the early steps, the hand shape tends to be quite distorted, and it only becomes more natural in the later steps.  
But in your case, the visualization seems to suggest that you don’t need many diffusion iterations. When I previously tested with a smaller number of diffusion steps, the overall success rate dropped a lot.  

If you have any further questions, I’d be very happy to discuss them with you.

**Comment by ShaoxiongYao** [comment]
Thank you for the quick response. I later realized I was visualizing the predicted initial config \hat{h}_0(h_t) in the GraspWithObject class. This is due to my misunderstanding of Equation 13, where the physics-aware loss terms should be evaluated at the predicted initial config instead of the current config. 

<img width="780" height="182" alt="Image" src="https://github.com/user-attachments/assets/f22e49b4-ab1d-4bf9-ac35-05c22a04751c" />

I will keep this issue open for a bit and let me know if I have further questions.
